libs/contrib/Test/Golden.idr

```diff
@@ -0,0 +1,194 @@
+         isSuccess <- runTest'
+         changeDir "../.."
+         pure isSuccess
+    where
+        getAnswer : IO Bool
+        getAnswer = do
+          str <- getLine
+          case str of
+            "y" => pure True
+            "n" => pure False
+            _   => do putStrLn "Invalid Answer."
+                      getAnswer
+
+        printExpectedVsOutput : String -> String -> IO ()
+        printExpectedVsOutput exp out = do
+          putStrLn "Expected:"
+          printLn exp
+          putStrLn "Given:"
+          printLn out
+
+        mayOverwrite : Maybe String -> String -> IO ()
+        mayOverwrite mexp out = do
+          the (IO ()) $ case mexp of
+            Nothing => putStr $ unlines
+              [ "Golden value missing. I computed the following result:"
+              , out
+              , "Accept new golden value? [yn]"
+              ]
+            Just exp => do
+              putStrLn "Golden value differs from actual value."
+              code <- system "git diff --no-index --exit-code --word-diff=color expected output"
+              when (code < 0) $ printExpectedVsOutput exp out
+              putStrLn "Accept actual value as new golden value? [yn]"
+          b <- getAnswer
+          when b $ do Right _ <- writeFile "expected" out
+                          | Left err => print err
+                      pure ()
+        runTest' : IO Bool
+        runTest'
+            = do putStr $ testPath ++ ": "
+                 system $ "sh ./run " ++ idris2 opts ++ " | tr -d '\\r' > output"
+                 Right out <- readFile "output"
+                     | Left err => do print err
+                                      pure False
+                 Right exp <- readFile "expected"
+                     | Left FileNotFound => do
+                         if interactive opts
+                           then mayOverwrite Nothing out
+                           else print FileNotFound
+                         pure False
+                     | Left err => do print err
+                                      pure False
+                 let result = normalize out == normalize exp
+                 if result
+                    then putStrLn "success"
+                    else do
+                      putStrLn "FAILURE"
+                      if interactive opts
+                         then mayOverwrite (Just exp) out
+                         else printExpectedVsOutput exp out
+
+                 pure result
+
+||| Check if a file exists for reading.
+export
+exists : String -> IO Bool
+exists f
+    = do Right ok <- openFile f Read
+             | Left err => pure False
+         closeFile ok
+         pure True
+
+firstExists : List String -> IO (Maybe String)
+firstExists [] = pure Nothing
+firstExists (x :: xs) = if !(exists x) then pure (Just x) else firstExists xs
+
+||| Find the first occurance of an executable on `PATH`.
+export
+pathLookup : List String -> IO (Maybe String)
```

Comment:
These functions are not used in the library. And neither is `onlyNames`.
Perhaps we need a function that abstract away the job done by `main` in
Idris' test runner?

A `TestTree` could be:

* a list of requirements (e.g. the `"CHEZ"`, `"NODE"`, etc. dependencies)
* a list of FileNames

and the golden runner would take a `List TestTree` and for each one of them:

* check whether filtering by `onlyNames` leaves any test
* if so then check the requirements are available
* if so then traverse the list of filtered tests with `runTest`

Reply:
Yes, I do not disagree but all this can be done in a later PR.

Reply:
I can add the changes if you want. It's so small (and justifies the presence
of currently unused features) it feels silly to split it in a different PR
altogether.

Reply:
go ahead!